### PR TITLE
=cluster handle aggressively rejoining/replacing nodes from same host/port pair

### PR DESCRIPTION
**Pull request by ktoso**
Resolves : https://github.com/apple/swift-distributed-actors/issues/1082

----

Short version: there is an edge case that was not handled well when very aggressively rejoining the cluster from the same host/port.

This can happen in k8s when a pod gets aggressively restarted, or on command line apps when someone joins a cluster "sends just one request and kills the app" since they both then still may be present in SWIM gossip (correctly) as dead, and the cluster may misinterpret this about information about "itself" when the new node joins.

----

More analysis:

steps:

- 7 joins 8
- 7 is leader
  - minumum 2 nodes before we elect one
  - lower address wins
- 7 dies
- 8 cannot declare 7 as down
  - only leader can do this
  - this is ok, as designed // these systems expect tog et back their node count and then recover
- 7 reboot -- let's call it 77
  - same host/port
  - new UID
- handshake with 8
  - 8 accepts
  - 77 gets accept
  - 8 declares "previous 7" as down, since 77 is the replacement
- 8 declaring 7 down is correct
  - but it means we have a down 7 in membership
  - this is also correct; other nodes may not yet know about this, so we want to spread this down information that 8 first noticed
- gossip in includes old node 7 (okey)
  - Node 77 receives gossip through SWIM and that includes **7**:
  - `2022-11-01T13:08:56+0900 trace Client : actor/id=/user/swim actor/path=/user/swim cluster/node=sact://REPLACEMENT_77@127.0.0.1:7337 swim/incarnation=0 swim/members/all=["SWIM.Member(SWIMActor(id:sact://RemoteCluster:7602583950674506995@127.0.0.1:8337/user/swim, node:sact://RemoteCluster:7602583950674506995@127.0.0.1:8337, alive(incarnation: 0), protocolPeriod: 1)", "SWIM.Member(SWIMActor(id:/user/swim, node:sact://REPLACEMENT_77@127.0.0.1:7337, alive(incarnation: 0), protocolPeriod: 0)"] swim/members/count=2 swim/ping/origin=sact://RemoteCluster:7602583950674506995@127.0.0.1:8337/user/swim swim/ping/payload=membership([SWIM.Member(SWIMActor(id:sact://OLD_NODE_7@127.0.0.1:7337/user/swim, node:sact://OLD_NODE_7@127.0.0.1:7337, suspect(incarnation: 0, suspectedBy: Set([sact://sact@127.0.0.1:8337#7602583950674506995])), protocolPeriod: 56), SWIM.Member(SWIMActor(id:sact://RemoteCluster:7602583950674506995@127.0.0.1:8337/user/swim, node:sact://RemoteCluster:7602583950674506995@127.0.0.1:8337, alive(incarnation: 0), protocolPeriod: 0), SWIM.Member(SWIMActor(id:/user/swim, node:sact://REPLACEMENT_77@127.0.0.1:7337, alive(incarnation: 0), protocolPeriod: 56)]) swim/ping/seqNr=4 swim/protocolPeriod=1 swim/suspects/count=0 swim/timeoutSuspectsBeforePeriodMax=11 swim/timeoutSuspectsBeforePeriodMin=4 [DistributedCluster] Received ping@4`
  - Note this: `swim/ping/payload=membership([`
    - **`    SWIM.Member(SWIMActor(id:sact://OLD_NODE_7@127.0.0.1:7337/user/swim, node:sact://OLD_NODE_7@127.0.0.1:7337, suspect(incarnation: 0, suspectedBy: Set([sact://sact@127.0.0.1:8337#7602583950674506995])), protocolPeriod: 56),`**
    - **`SWIM.Member(SWIMActor(id:sact://REPLACEMENT_77@127.0.0.1:7337/user/swim, node:sact://REPLACEMENT_77@127.0.0.1:7337, alive(incarnation: 0), protocolPeriod: 56),`**
    - `SWIM.Member(SWIMActor(id:sact://RemoteCluster:7602583950674506995@127.0.0.1:8337/user/swim, node:sact://RemoteCluster:7602583950674506995@127.0.0.1:8337, alive(incarnation: 0), protocolPeriod: 0)`
  - In other words, SWIM spreads information about both nodes since it is not `confirmDead` yet -- THIS IS OK. But continuing to act on the removed node's information is NOT ok.

Long story short: SWIM tells us that a node on this address was dead, but we know we are not dead -- this should only happen on high level gossip, when we see a .down somewhere about us. So we can ignore this from the SWIM level.

This should also get fixed in SWIM itself though, I'll follow up there.

### CODE REVIEW

File: Sources/DistributedActorsTestKit/LogCapture.swift [hunk 1]
```diff
@@ -115,10 +115,13 @@ extension LogCapture {
...
-            var actorPath: String = ""
+            var actorIdentifier: String = ""
             if var metadata = log.metadata {
-                if let path = metadata.removeValue(forKey: "actor/path") {
-                    actorPath = "[\(path)]"
+                if let id = metadata.removeValue(forKey: "actor/id") {
+                    actorIdentifier = "[\(id)]"
+                    _ = metadata.removeValue(forKey: "actor/path") // discard it
+                } else if let path = metadata.removeValue(forKey: "actor/path") {
+                    actorIdentifier = "[\(path)]"
```
Comment (ktoso) [inline Sources/DistributedActorsTestKit/LogCapture.swift:124]: revamping this now that we have more `distributed actor` types and less paths 👍

File: Samples/Sources/SampleDiningPhilosophers/SamplePrettyLogHandler.swift [hunk 2]
```diff
@@ -106,7 +106,7 @@ struct SamplePrettyLogHandler: LogHandler {
 
         let file = file.split(separator: "/").last ?? ""
         let line = line
-        print("\(self.timestamp()) [\(file):\(line)] [\(nodeInfo)\(Self.CONSOLE_BOLD)\(label)\(Self.CONSOLE_RESET)] [\(level)] \(message)\(metadataString)")
+        print("\(self.timestamp()) \(level) [\(nodeInfo)\(Self.CONSOLE_BOLD)\(label)\(Self.CONSOLE_RESET)][\(file):\(line)] \(message)\(metadataString)")
```
Comment (ktoso) [inline Samples/Sources/SampleDiningPhilosophers/SamplePrettyLogHandler.swift:109]: same formatting as our example formatters in samples

File: Sources/DistributedCluster/Cluster/Cluster+Membership.swift [hunk 3]
```diff
@@ -90,14 +90,22 @@ extension Cluster {
...
                 .count
         }
 
+        /// More efficient than using ``members(atMost:)`` followed by a `.count`
+        public func count(atMost status: Cluster.MemberStatus) -> Int {
+            self._members.values
+                .lazy
+                .filter { member in member.status <= status }
+                .count
+        }
```
Comment (ktoso) [inline Sources/DistributedCluster/Cluster/Cluster+Membership.swift:107]: used in test but generally useful

File: Sources/DistributedCluster/Cluster/Cluster+Membership.swift [hunk 4]
```diff
@@ -223,6 +238,21 @@ extension Cluster {
     }
 }
 
+extension Cluster.Membership: Sequence {
```
Comment (ktoso) [inline Sources/DistributedCluster/Cluster/Cluster+Membership.swift:241]: generally useful to `for member in membership` 👍

File: Sources/DistributedCluster/Cluster/SWIM/SWIMActor.swift [hunk 5]
```diff
@@ -355,6 +355,13 @@ internal distributed actor SWIMActor: SWIMPeer, SWIMAddressablePeer, CustomStrin
             )
         }
 
+        /// If SWIM claims we are dead, ignore this; we should be informed about this in high-level gossip soon enough.
+        if change.status == .dead,
+           change.member.node.asClusterNode == self.id.node
+        {
+            return
+        }
```
Comment (ktoso) [inline Sources/DistributedCluster/Cluster/SWIM/SWIMActor.swift:363]: The fix™

Though this deserves a deeper dig into SWIM as well.

Also, SwiftFormat's formatting of multi-line if is annoying omg!

File: Sources/DistributedCluster/Cluster/Cluster+Membership.swift [hunk 6]
```diff
@@ -153,6 +161,13 @@ extension Cluster {
             }
         }
 
+        /// Returns all members that are part of this membership, and have the any ``Cluster/MemberStatus`` that is *at most*
+        /// the passed in `status` passed in and `reachability` status. See ``Cluster/MemberStatus`` to learn more about the meaning of "at most".
+        ///
+        /// - Parameters:
+        ///   - statuses: statuses for which to check the members for
+        ///   - reachability: optional reachability that is the members will be filtered by
+        /// - Returns: array of members matching those checks. Can be empty.
```
Comment (yim-lee) [inline Sources/DistributedCluster/Cluster/Cluster+Membership.swift:170]:
```suggestion
        /// Returns all members that are part of this membership, and have ``Cluster/MemberStatus`` that is *at most*
        /// the passed in `status` and `reachability`. See ``Cluster/MemberStatus`` to learn more about the meaning of "at most".
        ///
        /// - Parameters:
        ///   - status: "at most" status for which to check the members for
        ///   - reachability: optional reachability that the members will be filtered by
        /// - Returns: array of members matching those checks. Can be empty.
```

File: Sources/MultiNodeTestKit/MultiNodeTestKit+Control.swift [hunk 7]
```diff
@@ -75,6 +93,7 @@ extension MultiNodeTest {
 // MARK: Run pieces of code on specific node
 
 extension MultiNodeTest.Control {
+    /// Runs a piece of code only on the specified `node`.
```
Comment (yim-lee) [inline Sources/MultiNodeTestKit/MultiNodeTestKit+Control.swift:96]: 🤔 the comment doesn't seem to fit the code?

**Comment by ktoso** [comment]
SWIM follow up https://github.com/apple/swift-cluster-membership/issues/91

File: Sources/MultiNodeTestKit/MultiNodeTestKit+Control.swift (hunk 7 above)
Comment (ktoso) [reply Sources/MultiNodeTestKit/MultiNodeTestKit+Control.swift:96]: Oh right needs a small rewording, on it

Comment (ktoso) [reply Sources/MultiNodeTestKit/MultiNodeTestKit+Control.swift:96]:
```suggestion
    /// Returns `true` if the multi-node execution is currently running on the same
    /// `node` as is passed to this function. This allows for executing pieces of code, 
    /// only on a specific node in the multi-node test execution.
```

**Comment by ktoso** [comment]
How I love a trailing space being the only failure :P
